galsim/chromatic.py

```diff
@@ -1359,3 +1376,286 @@ def withScaledFlux(self, flux_ratio):
+
+    def withSED(self, sed='flat'):
+        """
+        Make a new InterpolatedChromaticObject with the requested SED.
+
+        @param sed    Either a galsim.SED object to assign to this object as its SED, or a string
+                      'flat' which means a flat SED (equal weight to all wavelengths).  'flat' is
+                      what should be used for PSFs, so they do not modify the flux of the objects
+                      with which they are being convolved. 
+                      [default: 'flat']
+        @return a new InterpolatedChromaticObject
+        """
+        new = self.copy()
+        if sed == 'flat':
+            new.SED = lambda w: 1.0
+        else:
+            new.SED = sed
+        return new
+
+    def evaluateAtWavelength(self, wave, force_eval=False):
+        """
+        Evaluate this InterpolatedChromaticObject at a particular wavelength, either using
+        interpolation or via direct calculation, depending on how the object was originally
+        instantiated.
+
+        @param wave     Wavelength in nanometers.
+
+        @returns the monochromatic object at the given wavelength, as an InterpolatedImage.
+        """
+        if self.waves is not None and not force_eval:
+            im, stepk, maxk = self._imageAtWavelength(wave)
+            return galsim.InterpolatedImage(im, _force_stepk=stepk, _force_maxk=maxk)
+        else:
+            return self.simpleEvaluateAtWavelength(wave)
+
+    def _imageAtWavelength(self, wave):
+        """
+        Get an image of the object at a particular wavelength, using linear interpolation between
+        the originally-stored images.
+        """
+        # First, some wavelength-related sanity checks.
+        if self.waves is None:
+            raise RuntimeError("Requested image at some wavelength when doing direct calculation!")
+        if wave < min(self.waves) or wave > max(self.waves):
+            raise RuntimeError("Requested wavelength %.1f is outside the allowed range:"
+                               " %.1f to %.1f nm"%(wave, min(self.waves), max(self.waves)))
+
+        # Figure out where the supplied wavelength is compared to the list of wavelengths on which
+        # images were originally tabulated.
+        lower_idx = np.searchsorted(self.waves, wave)-1
+        # There can be edge issues, so watch out for that:
+        if lower_idx < 0: lower_idx = 0
+        if lower_idx == len(self.waves)-1: lower_idx = len(self.waves)-1
+        frac = (wave-self.waves[lower_idx]) / (self.waves[lower_idx+1]-self.waves[lower_idx])
+
+        # Actually do the interpolation for the image, stepK, and maxK.
+        im = _linearInterp(self.ims, frac, lower_idx)
+        stepk = _linearInterp(self.stepK_vals, frac, lower_idx)
+        maxk = _linearInterp(self.maxK_vals, frac, lower_idx)
+        return self.SED(wave)*im, stepk, maxk
+
+    def drawImage(self, bandpass, force_eval=False, image=None, **kwargs):
```

Comment:
I think any polynomial interpolation rule will have this property that there are at most ```len(self.waves)``` distinct terms that are being summed, with well-defined coefficients.

To be honest, the reason I did it this way was that the code seemed straightforward, and I thought there was no way that this was the limiting step.  Instantiation of InterpolatedImages seemed like it should be more expensive.  To test out this gut feeling, I just wrote a script that draws the ChromaticOpticalPSF from the unit tests, but with differently thinned versions of the LSST r bandpass that have a different ```len(bandpass.wave_list)```.  For ```len(bandpass.wave_list)``` = 23, 54, 73, the time to render a PSF image (not convolved with galaxy) was typically 0.5, 0.75, and 1.7 seconds.  That is a kind of weird scaling that I'm not quite sure how to interpret - I'm guessing its the sum of a nearly constant overhead (for the instantiation of the InterpolatedImage that always happens before drawing) plus a term that scales with the length of the combined wave list.  But anyway, it seems you're right - this is more important than I was expecting and I think I should refactor the code to account for it.

The way this would work (I think) is that I would have a subroutine that does the steps in ```_imageAtWavelength``` that find ```lower_idx``` and ```frac``` to interpolate between two stored images.  Using that subroutine here as well, I could figure out (for each wavelength of the combined wave list) which of the stored images contribute to the integral, and with what relative weights.  Then I could just take the total weighted sum, without calling any kind of ```DirectImageIntegrator``` at all.  (I suppose I could leave that new class just in case someone wants to implement a version of the integration that doesn't just use linear/polynomial interpolation between the stored images. But I don't need it if I do what you suggested, I think.)